aries_cloudagent/protocols/issue_credential/v3_0/formats/ld_proof/handler.py

```diff
@@ -0,0 +1,628 @@
+"""V2.0 issue-credential linked data proof credential format handler."""
+
+
+from ......vc.ld_proofs.error import LinkedDataProofException
+from ......vc.ld_proofs.check import get_properties_without_context
+import logging
+
+from typing import Mapping
+
+from marshmallow import EXCLUDE, INCLUDE
+
+from pyld import jsonld
+from pyld.jsonld import JsonLdProcessor
+
+from ......did.did_key import DIDKey
+from ......messaging.decorators.attach_decorator import AttachDecorator
+from ......storage.vc_holder.base import VCHolder
+from ......storage.vc_holder.vc_record import VCRecord
+from ......vc.vc_ld import (
+    issue_vc as issue,
+    verify_credential,
+    VerifiableCredentialSchema,
+    LDProof,
+    VerifiableCredential,
+)
+from ......vc.ld_proofs import (
+    AuthenticationProofPurpose,
+    BbsBlsSignature2020,
+    CredentialIssuancePurpose,
+    DocumentLoader,
+    Ed25519Signature2018,
+    LinkedDataProof,
+    ProofPurpose,
+    WalletKeyPair,
+)
+from ......vc.ld_proofs.constants import SECURITY_CONTEXT_BBS_URL
+from ......wallet.base import BaseWallet, DIDInfo
+from ......wallet.error import WalletNotFoundError
+from ......wallet.key_type import BLS12381G2, ED25519
+
+from ...message_types import (
+    ATTACHMENT_FORMAT,
+    CRED_30_ISSUE,
+    CRED_30_OFFER,
+    CRED_30_PROPOSAL,
+    CRED_30_REQUEST,
+)
+from ...messages.cred_format import V30CredFormat
+from ...messages.cred_issue import V30CredIssue
+from ...messages.cred_offer import V30CredOffer
+from ...messages.cred_proposal import V30CredProposal
+from ...messages.cred_request import V30CredRequest
+from ...models.cred_ex_record import V30CredExRecord
+from ...models.detail.ld_proof import V30CredExRecordLDProof
+
+from ..handler import CredFormatAttachment, V30CredFormatError, V30CredFormatHandler
+
+from .models.cred_detail import LDProofVCDetailSchema
+from .models.cred_detail import LDProofVCDetail
+
+LOGGER = logging.getLogger(__name__)
+
+SUPPORTED_ISSUANCE_PROOF_PURPOSES = {
+    CredentialIssuancePurpose.term,
+    AuthenticationProofPurpose.term,
+}
+SUPPORTED_ISSUANCE_SUITES = {Ed25519Signature2018}
+SIGNATURE_SUITE_KEY_TYPE_MAPPING = {Ed25519Signature2018: ED25519}
+
+
+# We only want to add bbs suites to supported if the module is installed
+if BbsBlsSignature2020.BBS_SUPPORTED:
+    SUPPORTED_ISSUANCE_SUITES.add(BbsBlsSignature2020)
+    SIGNATURE_SUITE_KEY_TYPE_MAPPING[BbsBlsSignature2020] = BLS12381G2
+
+
+PROOF_TYPE_SIGNATURE_SUITE_MAPPING = {
+    suite.signature_type: suite for suite in SIGNATURE_SUITE_KEY_TYPE_MAPPING
+}
+
+
+KEY_TYPE_SIGNATURE_SUITE_MAPPING = {
+    key_type: suite for suite, key_type in SIGNATURE_SUITE_KEY_TYPE_MAPPING.items()
+}
+
+
+class LDProofCredFormatHandler(V30CredFormatHandler):
```

Comment:
In AFJ we have the following model:
1. Credentials Api provides interface to issue credentials. You provide a protocol version which indicates the protocol version, but the api is the same otherwise for all versions
2. Credential Services are implemented for the different protocol version. The v1 protocol version has a hard depenency on the indy format and will throw an error if you try to issue a format that is other than indy. 
3. Credential Format Service provide the format specific implementation. This can be used by the credential services of different versions. The indy credential format service is currently being used by the v1 and v2 credential service and once the v3 credential service is added it will also leverage the indy format. This means the indy format is reused as much as possible and we get consistent behaviour whether you're using v1, v2 or v3.

Not sure if this is the approach ACA-Py wants to take, but I'm a bit worried about copying everything over when a new version is added instead of creating reusable abstract classes and interfaces.

Reply:
Thank you very much for your comment @TimoGlastra.

We went for the approach of creating the v3 folder after the decision in this meeting (the relevant minutes of the recording of the meeting are 15:45-21:40): 

https://wiki.hyperledger.org/display/ARIES/2022-04-19+Aries+Cloud+Agent+-+Python+Users+Group+Community+Meeting?preview=/62249226/62249966/20220419%20ACA-Pug%20Community%20Call%20Recording.mp4

Maybe someone of the ACA-Py community could also have a look at the PR because I am not sure how to proceed now (next step would be implementing Out-of-Band for DIDComm v2).

Reply:
I think the discussion here is about the present proof v2 vs the present proof v3 implementation. These protocols are indeed different and I see the need for a separate implementation. However the attachment formats are separate RFCs  0592, 0593 & 0510. 

The attachments have been extracted out of the issue credential / present proof protocols, with one of the reasons being that they could be versioned separately (i.e. a change to the issue credential protocol, doesn't require the attachment to be changed, and vice versa). Thus, the implementation for the format whether you're using v2 or v3 of the protocol is exactly the same. If a new format version is introduced it will require a new format implementation, but both v2 and v3 of the protocol could leverage that new version format. 

For me it's on the same level that we don't have a separate implementation of the connection protocol whether we're using issue credential or present proof protocol. Connections are a separate RFC with a separate implementation that can be reused by other parts of ACA-Py. I see the same for attachments formats. You implement them once and they can be reused by different credential protocol versions.